src/class-bulk-task.php

```diff
@@ -0,0 +1,238 @@
+	public function filter__posts_where( $where, $query ) {
+		if ( spl_object_hash( $query ) === $this->object_hash ) {
+			global $wpdb;
+
+			return sprintf(
+				'AND %s.ID > %d AND %s.ID <= %d %s',
+				$wpdb->posts,
+				$this->min_id,
+				$wpdb->posts,
+				$this->min_id + self::STEPPING,
+				$where
+			);
+		}
+
+		return $where;
+	}
+
+	/**
+	 * Loop through any number of posts efficiently with a callback, and output
+	 * the progress.
+	 *
+	 * @param array    $args {
+	 *     WP_Query args. Some have overridden defaults, and some are fixed.
+	 *     Anything not mentioned below will operate as normal.
+	 *
+	 *     @type bool   $ignore_sticky_posts Always true.
+	 *     @type bool   $no_found_rows       Always true.
+	 *     @type string $order               Always 'ASC'.
+	 *     @type string $orderby             Always 'ID'.
+	 *     @type int    $paged               Always 1.
+	 *     @type string $post_status         Defaults to 'any'.
+	 *     @type string $post_type           Defaults to 'any'.
+	 *     @type int    $posts_per_page      Defaults to 100.
+	 *     @type bool   $suppress_filters    Always false.
+	 * }
+	 * @param callable $callable Callback function to invoke for each post.
+	 *                           The callable will be passed a post object.
+	 */
+	public function run( array $args, callable $callable ): void {
+		global $wpdb;
+
+		// Apply default arguments.
+		$args = wp_parse_args(
+			$args,
+			[
+				'post_type'      => 'any',
+				'post_status'    => 'any',
+				'posts_per_page' => 100,
+			],
+		);
+
+		// Force some arguments and don't let them get overridden.
+		$args['ignore_sticky_posts'] = true;
+		$args['no_found_rows']       = true;
+		$args['order']               = 'ASC';
+		$args['orderby']             = 'ID';
+		$args['paged']               = 1;
+		$args['suppress_filters']    = false;
+
+		// Set the min ID from the cursor.
+		$this->min_id = $this->cursor->get();
+
+		// Set the max ID from the database.
+		$this->max_id = $wpdb->get_var( 'SELECT MAX(ID) FROM ' . $wpdb->posts ); // phpcs:ignore WordPress.DB.DirectDatabaseQuery.DirectQuery, WordPress.DB.DirectDatabaseQuery.NoCaching
+
+		// Handle pagination.
+		add_filter( 'posts_where', [ $this, 'filter__posts_where' ], 9999, 2 );
+
+		// Turn off some automatic behavior that would slow down the process.
+		$this->before_run();
+
+		// All systems go.
+		while ( $this->min_id < $this->max_id ) {
+			// Build the query object, but don't run it without the object hash.
+			$query = new WP_Query();
+
+			// Store the unique object hash to ensure we only filter this query.
+			$this->object_hash = spl_object_hash( $query );
+
+			// Run the query.
+			$query->query( $args );
+
+			// Fork for results vs. not.
+			if ( $query->have_posts() ) {
+				// Invoke the callable over every post.
+				array_walk( $query->posts, $callable );
+
+				// Update our min ID for the next query.
+				$this->min_id = max( wp_list_pluck( $query->posts, 'ID' ) );
```

Comment:
I know this is legacy, but I don't recall why I did it that way. Seems like we should be able to take the ID of the last post instead... right?

Reply:
Yep, we should be able to.